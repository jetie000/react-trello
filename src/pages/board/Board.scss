@import '../../variables.scss';

.board-info {
    width: 350px;
    height: calc(100vh - 85px);

    &>:first-child {
        height: 250px;
    }

    &>:nth-child(2) {
        height: calc(100vh - 350px);
        &>h6{
            min-height: 100px;
        }
    }
}


.board-users-avatar {
    width: 50px;
    height: 50px;
}

.board-users-avatar-sm{
    width: 25px;
    height: 25px;
}

.board-columns {
    height: calc(100vh - 85px);

    &>div:first-child {

        align-items: flex-start;

        &>* {
            width: 200px;
            overflow-y: auto;
            max-height: calc(100vh - 125px);
        }

        &>button{
            height: 100px;
        }
    }
}

.board-tasks{
    &>div{
        height: 200px;
    }
}
