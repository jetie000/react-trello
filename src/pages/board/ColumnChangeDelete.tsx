import Modal from "@/components/modal/Modal"
import { useChangeColumnMutation, useDeleteColumnMutation } from "@/store/api/column.api"
import { IColumn } from "@/types/column.interface"
import * as React from "react"
import { useRef } from "react"
import { Toast as bootstrapToast } from "bootstrap"
import { Modal as bootstrapModal } from "bootstrap"
import { useActions } from "@/hooks/useActions"
import { variables } from "@/variables"
import { useSelector } from "react-redux"
import { RootState } from "@/store/store"

function ColumnChangeDelete({ currentColumn }: { currentColumn: IColumn | undefined }) {
  const { showToast } = useActions()
  const { language } = useSelector((state: RootState) => state.options)
  const modalRefDelete = useRef<HTMLDivElement>(null)
  const modalRefChange = useRef<HTMLDivElement>(null)

  const changeColumnRef = useRef<HTMLInputElement>(null)
  const [
    changeColumn,
    { isSuccess: isSuccessChange, isError: isErrorChange, isLoading: isLoadingChange }
  ] = useChangeColumnMutation()
  const [
    deleteColumn,
    { isSuccess: isSuccessDelete, isError: isErrorDelete, isLoading: isLoadingDelete }
  ] = useDeleteColumnMutation()

  const addColumnClick = () => {
    if (changeColumnRef.current && changeColumnRef.current.value !== "" && currentColumn) {
      changeColumn({
        id: currentColumn.id,
        name: changeColumnRef.current.value,
        order: currentColumn.order
      })
    }
  }

  React.useEffect(() => {
    if (isSuccessDelete) {
      if (modalRefDelete.current) {
        const myModal = bootstrapModal.getOrCreateInstance("#" + modalRefDelete.current?.id)
        myModal.hide()
      }
      showToast(variables.LANGUAGES[language].COLUMN_DELETED)
    }
    if (isErrorDelete) {
      showToast(variables.LANGUAGES[language].ERROR_REQUEST)
    }
  }, [isLoadingDelete])

  React.useEffect(() => {
    if (isSuccessChange) {
      showToast(variables.LANGUAGES[language].COLUMN_CHANGED)
    }
    if (isErrorChange) {
      showToast(variables.LANGUAGES[language].ERROR_REQUEST)
    }
  }, [isLoadingChange])

  return (
    <>
      <Modal
        title={variables.LANGUAGES[language].CHANGE_COLUMN}
        id="changeColumn"
        size="sm"
        ref={modalRefChange}
      >
        <div className="d-flex flex-column">
          <label htmlFor="inputColumnNameChange">{variables.LANGUAGES[language].NAME}</label>
          <input
            className="form-control mb-2"
            id="inputColumnNameChange"
            placeholder={variables.LANGUAGES[language].ENTER_NAME}
            ref={changeColumnRef}
            defaultValue={currentColumn?.name}
          />
          <button className="btn btn-primary" onClick={addColumnClick}>
            {variables.LANGUAGES[language].CHANGE_COLUMN}
          </button>
        </div>
      </Modal>
      <Modal
        id="deleteColumn"
        title={variables.LANGUAGES[language].DELETE_COLUMN}
        size="sm"
        ref={modalRefDelete}
      >
        <div className="d-flex flex-column gap-2">
          <div>{variables.LANGUAGES[language].SURE_DELETE_COLUMN}</div>
          <button className="btn btn-danger" onClick={() => deleteColumn(currentColumn?.id || 0)}>
            {variables.LANGUAGES[language].DELETE_COLUMN}
          </button>
        </div>
      </Modal>
    </>
  )
}

export default ColumnChangeDelete
