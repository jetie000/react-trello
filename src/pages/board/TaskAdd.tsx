import UsersList from "@/components/usersList/UsersList"
import { useAddTaskMutation } from "@/store/api/task.api"
import { IColumn } from "@/types/column.interface"
import * as React from "react"
import { useRef, useState } from "react"
import { Toast as bootstrapToast } from "bootstrap"
import { Modal as bootstrapModal } from "bootstrap"
import { useActions } from "@/hooks/useActions"
import { RootState } from "@/store/store"
import { useSelector } from "react-redux"
import { variables } from "@/variables"

function TaskAdd({ column }: { column: IColumn | undefined }) {
  const { setToastChildren } = useActions()
  const { language } = useSelector((state: RootState) => state.options)

  const addTaskNameRef = useRef<HTMLInputElement>(null)
  const addTaskDescRef = useRef<HTMLTextAreaElement>(null)
  const [addTask, { isSuccess: isSuccessAdd, isError: isErrorAdd, isLoading: isLoadingAdd }] =
    useAddTaskMutation()
  const [userIds, setUserIds] = useState<number[]>([])

  const addTaskClick = () => {
    if (addTaskNameRef.current && addTaskNameRef.current.value !== "" && column) {
      addTask({
        columnId: column.id,
        name: addTaskNameRef.current.value,
        description: addTaskDescRef.current?.value,
        userIds: userIds
      })
    }
  }

  React.useEffect(() => {
    if (isSuccessAdd) {
      const myModal = bootstrapModal.getOrCreateInstance(
        document.getElementById("addTask") || "addTask"
      )
      myModal.hide()
      const myToast = bootstrapToast.getOrCreateInstance(
        document.getElementById("myToast") || "myToast"
      )
      setToastChildren(variables.LANGUAGES[language].TASK_ADDED)
      myToast.show()
    }
    if (isErrorAdd) {
      const myToast = bootstrapToast.getOrCreateInstance(
        document.getElementById("myToast") || "myToast"
      )
      setToastChildren(variables.LANGUAGES[language].ERROR_REQUEST)
      myToast.show()
    }
  }, [isLoadingAdd])

  return (
    <div className="d-flex flex-column">
      <label htmlFor="inputTaskName">{variables.LANGUAGES[language].NAME}</label>
      <input
        className="form-control mb-2"
        id="inputTaskName"
        placeholder={variables.LANGUAGES[language].ENTER_NAME}
        ref={addTaskNameRef}
      />
      <label htmlFor="inputTaskDesc">{variables.LANGUAGES[language].DESCRIPTION}</label>
      <textarea
        className="form-control mb-2"
        id="inputTaskDesc"
        placeholder={variables.LANGUAGES[language].ENTER_DESCRIPTION}
        ref={addTaskDescRef}
      />
      <UsersList userIds={userIds} setUserIds={setUserIds} boardId={column?.boardId} />
      <button className="btn btn-primary mt-2" onClick={addTaskClick}>
        {variables.LANGUAGES[language].ADD_TASK}
      </button>
    </div>
  )
}

export default TaskAdd
