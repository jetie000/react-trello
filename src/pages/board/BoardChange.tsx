import UsersList from "../../components/usersList/UsersList"
import { useActions } from "@/hooks/useActions"
import { useEffect, useRef, useState } from "react"
import { IBoard } from "@/types/board.interface"
import { useChangeBoardMutation } from "@/store/api/board.api"
import { RootStateStore } from "@/store/store"
import { useSelector } from "react-redux"
import { languages } from "@/config/languages"

function BoardChange({ board }: { board: IBoard }) {
  const { showToast } = useActions()

  const { language } = useSelector((state: RootStateStore) => state.options)
  const [userIds, setUserIds] = useState<number[]>(board.users?.map(u => u.id) || [])
  const nameRef = useRef<HTMLInputElement>(null)
  const descRef = useRef<HTMLTextAreaElement>(null)

  const [
    changeBoard,
    { isLoading: isLoadingChange, isError: isErrorChange, isSuccess: isSuccessChange }
  ] = useChangeBoardMutation()

  useEffect(() => {
    if (isErrorChange) {
      showToast(languages[language].ERROR_REQUEST)
    }
    if (isSuccessChange) {
      showToast(languages[language].BOARD_CHANGED)
    }
  }, [isLoadingChange])

  const changeBoardClick = () => {
    if (nameRef.current && descRef.current && nameRef.current?.value.trim() !== "") {
      changeBoard({
        name: nameRef.current.value,
        description: descRef.current.value,
        id: board.id,
        userIds
      })
    } else {
      showToast(languages[language].INPUT_DATA)
    }
  }

  return (
    <div>
      <div className="mb-3">
        <label htmlFor="inputBoardName">{languages[language].NAME}</label>
        <input
          className="form-control"
          id="inputBoardName"
          placeholder={languages[language].ENTER_NAME}
          ref={nameRef}
          defaultValue={board.name}
        />
      </div>
      <div className="mb-3">
        <label htmlFor="inputBoardDescription">{languages[language].DESCRIPTION}</label>
        <textarea
          className="form-control"
          id="inputBoardDescription"
          placeholder={languages[language].ENTER_DESCRIPTION}
          ref={descRef}
          defaultValue={board.description}
        />
      </div>
      <UsersList userIds={userIds} setUserIds={setUserIds} />
      <button
        type="button"
        className="btn btn-primary w-100 mt-3"
        onClick={changeBoardClick}
        data-testid="change-board-btn"
      >
        {isLoadingChange ? (
          <div className="spinner-border spinner-border-sm" role="status">
            <span className="visually-hidden">{languages[language].LOADING}</span>
          </div>
        ) : (
          languages[language].CHANGE_BOARD
        )}
      </button>
    </div>
  )
}

export default BoardChange
