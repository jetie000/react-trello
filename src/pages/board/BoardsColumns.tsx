import { IBoard } from '@/types/board.interface';
import * as React from 'react';


function BoardsColumns({board}: {board: IBoard}) {
    return ( 
        <div>

        </div>
     );
}

export default BoardsColumns;