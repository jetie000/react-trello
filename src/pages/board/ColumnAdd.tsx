import { useAddColumnMutation } from '@/store/api/column.api';
import { IBoard } from '@/types/board.interface';
import * as React from 'react';
import { Toast as bootstrapToast } from 'bootstrap';
import { Modal as bootstrapModal } from 'bootstrap';
import { useActions } from '@/hooks/useActions';

function ColumnAdd({board}:{board: IBoard | undefined}) {
    const addColumnRef = React.useRef<HTMLInputElement>(null)
    const [addColumn, { isSuccess: isSuccessAdd, isError: isErrorAdd, isLoading: isLoadingAdd }] = useAddColumnMutation()
    const { setToastChildren } = useActions();

    
    React.useEffect(() => {
        if (isSuccessAdd) {
            const myModal = bootstrapModal.getOrCreateInstance(document.getElementById('addColumn') || 'addColumn');
            myModal.hide();
            const myToast = bootstrapToast.getOrCreateInstance(document.getElementById('myToast') || 'myToast');
            setToastChildren("Column added succesfully")
            myToast.show()
        }
        if (isErrorAdd) {
            const myToast = bootstrapToast.getOrCreateInstance(document.getElementById('myToast') || 'myToast');
            setToastChildren("Error in column adding")
            myToast.show()
        }
    }, [isLoadingAdd])
    
    const addColumnClick = () => {
        if (addColumnRef.current &&
            addColumnRef.current.value !== '' &&
            board && 'id' in board) {
            addColumn({
                boardId: board.id,
                name: addColumnRef.current.value
            })
        }
    }

    return (
        <div className="d-flex flex-column">
            <label htmlFor="inputColumnName">Name</label>
            <input className="form-control mb-2" id="inputColumnName"
                placeholder="Enter column name" ref={addColumnRef} />
            <button className='btn btn-primary' onClick={addColumnClick}>
                Add column
            </button>
        </div>
    );
}

export default ColumnAdd;