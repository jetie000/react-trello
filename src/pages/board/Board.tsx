import {
  useDeleteBoardMutation,
  useGetBoardByIdQuery,
  useLeaveBoardMutation
} from "@/store/api/board.api"
import { useParams } from "react-router"
import { useEffect, useMemo, useRef } from "react"
import { Modal } from "bootstrap"
import Columns from "./Columns"
import { useSelector } from "react-redux"
import { Navigate, useNavigate } from "react-router-dom"
import { RootStateStore } from "@/store/store"
import ModalWrapper from "../../components/modalWrapper/ModalWrapper"
import { useActions } from "@/hooks/useActions"
import BoardChange from "./BoardChange"
import ColumnAdd from "./ColumnAdd"
import { languages } from "@/config/languages"
import "./Board.scss"

function Board() {
  const { token, id: idUser } = useSelector((state: RootStateStore) => state.user)

  if (!token) {
    return <Navigate to={"/login"} />
  }

  const { id } = useParams()
  const { showToast } = useActions()
  const { language } = useSelector((state: RootStateStore) => state.options)
  const navigate = useNavigate()
  const modalRefDelete = useRef<HTMLDivElement>(null)
  const modalRefLeave = useRef<HTMLDivElement>(null)
  const modalRefChange = useRef<HTMLDivElement>(null)
  const { isLoading, isError, isSuccess, data } = useGetBoardByIdQuery(Number(id))
  const [
    deleteBoard,
    { isSuccess: isSuccessDelete, isError: isErrorDelete, isLoading: isLoadingDelete }
  ] = useDeleteBoardMutation()
  const [
    leaveBoard,
    { isSuccess: isSuccessLeave, isError: isErrorLeave, isLoading: isLoadingLeave }
  ] = useLeaveBoardMutation()

  const creatorName = useMemo(
    () => data && "name" in data && data.users?.find(u => u.id === data.creatorId)?.fullName,
    [data]
  )

  useEffect(() => {
    if (isSuccess && !data) navigate("/")
  }, [isLoading])

  useEffect(() => {
    if (isSuccessDelete) {
      if (modalRefDelete.current) {
        const myModal = Modal.getOrCreateInstance("#" + modalRefDelete.current?.id)
        myModal.hide()
      }
      navigate("/")
      showToast(languages[language].BOARD_DELETED)
    }
    if (isErrorDelete) {
      showToast(languages[language].ERROR_BOARD_DELETING)
    }
  }, [isLoadingDelete])

  useEffect(() => {
    if (isSuccessLeave) {
      navigate("/")
      showToast(languages[language].BOARD_LEFT)
    }
    if (isErrorLeave) {
      showToast(languages[language].ERROR_BOARD_LEAVING)
    }
  }, [isLoadingLeave])

  return isLoading ? (
    <div className="spinner-border m-auto" role="status">
      <span className="visually-hidden"> {languages[language].LOADING}</span>
    </div>
  ) : isError ? (
    <h1 className="m-auto">{languages[language].ERROR_REQUEST}</h1>
  ) : (
    data &&
    "name" in data && (
      <div className="d-flex gap-2 flex-fill min-w-0">
        <div className="d-flex flex-column board-info gap-2">
          <div className="d-flex flex-column border rounded-2 p-3 ">
            <h4 className="text-center">{data && data.name}</h4>
            <h6 className="text-center">
              {languages[language].CREATOR + ": "}
              {creatorName}
            </h6>
            {data?.description && <h6>{data.description}</h6>}
            {data.creatorId === idUser ? (
              <div className="d-flex gap-2 align-self-end mt-auto">
                <button
                  className="btn btn-primary"
                  data-bs-toggle="modal"
                  data-bs-target="#changeBoard"
                >
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    width="16"
                    height="16"
                    fill="currentColor"
                    className="bi bi-pencil-square"
                    viewBox="0 0 16 16"
                  >
                    <path d="M15.502 1.94a.5.5 0 0 1 0 .706L14.459 3.69l-2-2L13.502.646a.5.5 0 0 1 .707 0l1.293 1.293zm-1.75 2.456-2-2L4.939 9.21a.5.5 0 0 0-.121.196l-.805 2.414a.25.25 0 0 0 .316.316l2.414-.805a.5.5 0 0 0 .196-.12l6.813-6.814z" />
                    <path
                      fillRule="evenodd"
                      d="M1 13.5A1.5 1.5 0 0 0 2.5 15h11a1.5 1.5 0 0 0 1.5-1.5v-6a.5.5 0 0 0-1 0v6a.5.5 0 0 1-.5.5h-11a.5.5 0 0 1-.5-.5v-11a.5.5 0 0 1 .5-.5H9a.5.5 0 0 0 0-1H2.5A1.5 1.5 0 0 0 1 2.5z"
                    />
                  </svg>
                </button>
                <button
                  className="btn btn-danger"
                  data-bs-toggle="modal"
                  data-bs-target="#deleteBoard"
                >
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    width="16"
                    height="16"
                    fill="currentColor"
                    className="bi bi-trash3-fill"
                    viewBox="0 0 16 16"
                  >
                    <path d="M11 1.5v1h3.5a.5.5 0 0 1 0 1h-.538l-.853 10.66A2 2 0 0 1 11.115 16h-6.23a2 2 0 0 1-1.994-1.84L2.038 3.5H1.5a.5.5 0 0 1 0-1H5v-1A1.5 1.5 0 0 1 6.5 0h3A1.5 1.5 0 0 1 11 1.5m-5 0v1h4v-1a.5.5 0 0 0-.5-.5h-3a.5.5 0 0 0-.5.5M4.5 5.029l.5 8.5a.5.5 0 1 0 .998-.06l-.5-8.5a.5.5 0 1 0-.998.06m6.53-.528a.5.5 0 0 0-.528.47l-.5 8.5a.5.5 0 0 0 .998.058l.5-8.5a.5.5 0 0 0-.47-.528M8 4.5a.5.5 0 0 0-.5.5v8.5a.5.5 0 0 0 1 0V5a.5.5 0 0 0-.5-.5" />
                  </svg>
                </button>
              </div>
            ) : (
              <button
                className="btn btn-danger align-self-start mt-auto"
                data-bs-toggle="modal"
                data-bs-target="#leaveBoard"
              >
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  width="16"
                  height="16"
                  fill="currentColor"
                  className="bi bi-box-arrow-left"
                  viewBox="0 0 16 16"
                >
                  <path
                    fillRule="evenodd"
                    d="M6 12.5a.5.5 0 0 0 .5.5h8a.5.5 0 0 0 .5-.5v-9a.5.5 0 0 0-.5-.5h-8a.5.5 0 0 0-.5.5v2a.5.5 0 0 1-1 0v-2A1.5 1.5 0 0 1 6.5 2h8A1.5 1.5 0 0 1 16 3.5v9a1.5 1.5 0 0 1-1.5 1.5h-8A1.5 1.5 0 0 1 5 12.5v-2a.5.5 0 0 1 1 0z"
                  />
                  <path
                    fillRule="evenodd"
                    d="M.146 8.354a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L1.707 7.5H10.5a.5.5 0 0 1 0 1H1.707l2.147 2.146a.5.5 0 0 1-.708.708z"
                  />
                </svg>
              </button>
            )}
          </div>
          <div className="d-flex flex-column p-3 gap-3 overflow-y-auto flex-fill border rounded-2">
            {data.users?.map(user => (
              <div className="d-flex flex-column gap-3" key={user.id}>
                <div className="d-flex gap-3">
                  <div
                    className="border rounded-circle d-flex 
                                    align-items-center justify-content-center 
                                    bg-secondary text-light fs-4 flex-shrink-0 board-users-avatar"
                  >
                    {user.fullName[0]}
                  </div>
                  <div className="d-flex flex-column min-w-0">
                    <div className="text-truncate">{user.email}</div>
                    <div className="text-truncate">{user.fullName}</div>
                  </div>
                </div>
              </div>
            ))}
          </div>
        </div>
        <Columns boardColumns={data.columns} />
        <ColumnAdd boardId={data.id} />

        {data.creatorId === idUser ? (
          <>
            <ModalWrapper
              id="deleteBoard"
              title={languages[language].DELETE_BOARD}
              size="sm"
              ref={modalRefDelete}
            >
              <div className="d-flex flex-column gap-2">
                <div>{languages[language].SURE_DELETE_BOARD}</div>
                <button
                  className="btn btn-danger"
                  onClick={() => deleteBoard(data.id)}
                  data-testid="delete-board-btn"
                >
                  {languages[language].DELETE_BOARD}
                </button>
              </div>
            </ModalWrapper>
            <ModalWrapper
              id="changeBoard"
              title={languages[language].CHANGE_BOARD}
              size="md"
              ref={modalRefChange}
            >
              <BoardChange board={data} />
            </ModalWrapper>
          </>
        ) : (
          <ModalWrapper
            id="leaveBoard"
            title={languages[language].LEAVE_BOARD}
            size="sm"
            ref={modalRefLeave}
          >
            <div className="d-flex flex-column gap-2">
              <div>{languages[language].SURE_LEAVE_BOARD}</div>
              <button
                className="btn btn-danger"
                onClick={() => leaveBoard(data.id)}
                data-testid="leave-board-btn"
              >
                {languages[language].LEAVE_BOARD}
              </button>
            </div>
          </ModalWrapper>
        )}
      </div>
    )
  )
}

export default Board
