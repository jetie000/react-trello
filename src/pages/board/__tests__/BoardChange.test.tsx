import * as reduxHooks from "react-redux"
import * as actions from "@/hooks/useActions"
import * as boardApi from "@/store/api/board.api"
import * as userApi from "@/store/api/user.api"
import { fireEvent, render, screen } from "@testing-library/react"
import { languages } from "@/config/languages"
import { skipToken } from "@reduxjs/toolkit/query"
import { mockBoards, mockReturnUsers, mockUsers } from "../../../config/mockValues"
import { mockReturnMutation } from "@/config/mockValues"
import BoardChange from "../BoardChange"

jest.mock("@/store/store")
jest.mock("react-redux")
jest.mock("react-router-dom")
jest.mock("@/hooks/useActions")
jest.mock("@/store/api/user.api")
jest.mock("@/store/api/board.api")
const mockUseSelector = jest.spyOn(reduxHooks, "useSelector")
const mockUseActions = jest.spyOn(actions, "useActions")
const mockChange = jest.spyOn(boardApi, "useChangeBoardMutation")
const mockGetByIds = jest.spyOn(userApi, "useGetByIdsQuery")
const mockSearchUsers = jest.spyOn(userApi, "useSearchUsersQuery")

const mockReturnSelectorValue = {
  language: 1,
  id: 1,
  token: "token"
}
const mockChangeFunc = jest.fn()
const showToastMock = jest.fn()
const mockReturnActionsValue = { showToast: showToastMock }

const nameBoard = "Board 11"
const descriptionBoard = "Desc Board 11"

describe("BoardChange", () => {
  beforeEach(() => {
    jest.restoreAllMocks()
    jest.resetAllMocks()
    mockGetByIds.mockImplementation((ids: number[] | typeof skipToken) => {
      if (typeof ids !== "symbol")
        return mockReturnUsers(mockUsers.filter(u => ids.some(id => id === u.id))) as any
    })
    mockSearchUsers.mockReturnValue(mockReturnUsers(mockUsers))
    mockUseSelector.mockReturnValue(mockReturnSelectorValue)
    mockUseActions.mockReturnValue(mockReturnActionsValue as any)
    mockChange.mockReturnValue([mockChangeFunc, mockReturnMutation(false, false)])
  })
  it("should render component with all fields and buttons", () => {
    const component = render(<BoardChange board={mockBoards[0]} />)
    expect(
      screen.getByPlaceholderText(languages[mockReturnSelectorValue.language].ENTER_NAME)
    ).toBeInTheDocument()
    expect(
      screen.getByPlaceholderText(languages[mockReturnSelectorValue.language].ENTER_DESCRIPTION)
    ).toBeInTheDocument()
    expect(screen.getByTestId("change-board-btn")).toBeInTheDocument()
    expect(component).toMatchSnapshot()
  })
  it("should show enter data message when provided invalid data when changing board", () => {
    const component = render(<BoardChange board={mockBoards[0]} />)
    fireEvent.change(
      screen.getByPlaceholderText(languages[mockReturnSelectorValue.language].ENTER_NAME),
      {
        target: { value: "" }
      }
    )
    fireEvent.click(screen.getByTestId("change-board-btn"))
    expect(showToastMock).toHaveBeenCalledWith(
      languages[mockReturnSelectorValue.language].INPUT_DATA
    )
    expect(showToastMock).toHaveBeenCalledTimes(1)
  })
  it("should change task when all fields are filled correctly", () => {
    const component = render(<BoardChange board={mockBoards[0]} />)
    expect(mockUseActions).toHaveBeenCalled()
    expect(mockChange).toHaveBeenCalled()
    expect(mockUseSelector).toHaveBeenCalled()
    fireEvent.change(
      screen.getByPlaceholderText(languages[mockReturnSelectorValue.language].ENTER_NAME),
      {
        target: { value: nameBoard }
      }
    )
    fireEvent.change(
      screen.getByPlaceholderText(languages[mockReturnSelectorValue.language].ENTER_DESCRIPTION),
      {
        target: { value: descriptionBoard }
      }
    )
    fireEvent.click(screen.getByTestId("change-board-btn"))
    mockChange.mockReturnValue([mockChangeFunc, mockReturnMutation(true, true)])
    component.rerender(<BoardChange board={mockBoards[0]} />)
    mockChange.mockReturnValue([mockChangeFunc, mockReturnMutation(false, true)])
    component.rerender(<BoardChange board={mockBoards[0]} />)
    expect(mockChangeFunc).toHaveBeenCalledWith({
      id: mockBoards[0].id,
      name: nameBoard,
      description: descriptionBoard,
      userIds: mockBoards[0].users?.map(u => u.id) || []
    })
    expect(showToastMock).toHaveBeenCalledWith(
      languages[mockReturnSelectorValue.language].BOARD_CHANGED
    )
    expect(showToastMock).toHaveBeenCalledTimes(1)
  })
})
