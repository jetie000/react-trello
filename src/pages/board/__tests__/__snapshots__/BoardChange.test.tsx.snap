// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardChange should render component with all fields and buttons 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="mb-3"
        >
          <label
            for="inputBoardName"
          >
            Name
          </label>
          <input
            class="form-control"
            id="inputBoardName"
            placeholder="Input name"
            value="Board1"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            for="inputBoardDescription"
          >
            Description (optional)
          </label>
          <textarea
            class="form-control"
            id="inputBoardDescription"
            placeholder="Enter description"
          />
        </div>
        <div
          class="mb-1"
        >
          <label>
            Users
          </label>
          <div
            class="input-group"
          >
            <input
              class="form-control w-50"
              data-testid="search-users-input"
              placeholder="Enter request"
              type="text"
            />
            <button
              class="btn btn-secondary d-flex align-items-center add-board-behind-canvas"
              type="button"
            >
              <svg
                class="bi bi-x-lg"
                fill="currentColor"
                height="20"
                viewBox="0 0 16 16"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.146 2.854a.5.5 0 1 1 .708-.708L8 7.293l5.146-5.147a.5.5 0 0 1 .708.708L8.707 8l5.147 5.146a.5.5 0 0 1-.708.708L8 8.707l-5.146 5.147a.5.5 0 0 1-.708-.708L7.293 8z"
                />
              </svg>
            </button>
          </div>
        </div>
        <button
          class="btn btn-primary w-100 mt-3"
          data-testid="change-board-btn"
          type="button"
        >
          Change board
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="mb-3"
      >
        <label
          for="inputBoardName"
        >
          Name
        </label>
        <input
          class="form-control"
          id="inputBoardName"
          placeholder="Input name"
          value="Board1"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          for="inputBoardDescription"
        >
          Description (optional)
        </label>
        <textarea
          class="form-control"
          id="inputBoardDescription"
          placeholder="Enter description"
        />
      </div>
      <div
        class="mb-1"
      >
        <label>
          Users
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control w-50"
            data-testid="search-users-input"
            placeholder="Enter request"
            type="text"
          />
          <button
            class="btn btn-secondary d-flex align-items-center add-board-behind-canvas"
            type="button"
          >
            <svg
              class="bi bi-x-lg"
              fill="currentColor"
              height="20"
              viewBox="0 0 16 16"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.146 2.854a.5.5 0 1 1 .708-.708L8 7.293l5.146-5.147a.5.5 0 0 1 .708.708L8.707 8l5.147 5.146a.5.5 0 0 1-.708.708L8 8.707l-5.146 5.147a.5.5 0 0 1-.708-.708L7.293 8z"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="btn btn-primary w-100 mt-3"
        data-testid="change-board-btn"
        type="button"
      >
        Change board
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
